[Bug]: MACE fails to JIT compile with Python 3.10 union type hints

### Expected behavior

The MACE model uses `@torch.jit.script` for performance and would be good to maintain functionality.

### Actual behavior

The following error gets raised as I was writing example unit tests:

```console
    fn = torch._C._jit_script_compile(
E   RuntimeError: 
E   Expression of type | cannot be used in a type expression:
E     File "/home/kinlongk/Repos/matsciml/matsciml/models/pyg/mace/tools/scatter.py", line 39
E       index: torch.Tensor,
E       dim: int = -1,
E       out: torch.Tensor | None = None,
E            ~~~~~~~~~~~~~~~~~~~ <--- HERE
E       dim_size: int | None = None,
E       reduce: str = "sum",
```

According to [this PR](https://github.com/pytorch/pytorch/pull/109293), `@torch.jit.script` is in maintenance mode and does not support new Python features out of the box. The referenced PR is completed and merged, but will have to wait until PyTorch 2.2.0 to see it being available. 

The short term solution would be to annotate those lines to stop formatters from changing out the type hint to Python 3.10 unions.

### Steps to reproduce the problem

Related PyTorch issue: https://github.com/pytorch/pytorch/issues/114755

### Specifications

```console
absl-py==2.1.0
aiohttp==3.9.1
aiosignal==1.3.1
annotated-types==0.6.0
anyio @ file:///home/conda/feedstock_root/build_artifacts/anyio_1688651106312/work/dist
argon2-cffi @ file:///home/conda/feedstock_root/build_artifacts/argon2-cffi_1692818318753/work
argon2-cffi-bindings @ file:///home/conda/feedstock_root/build_artifacts/argon2-cffi-bindings_1695386548994/work
ase==3.22.1
asttokens @ file:///home/conda/feedstock_root/build_artifacts/asttokens_1698341106958/work
async-timeout==4.0.3
attrs==23.1.0
bandit==1.7.5
beautifulsoup4 @ file:///home/conda/feedstock_root/build_artifacts/beautifulsoup4_1705564648255/work
black==23.11.0
bleach @ file:///home/conda/feedstock_root/build_artifacts/bleach_1696630167146/work
Brotli @ file:///home/conda/feedstock_root/build_artifacts/brotli-split_1695989787169/work
cachetools==5.3.2
certifi @ file:///home/conda/feedstock_root/build_artifacts/certifi_1707022139797/work/certifi
cffi @ file:///home/conda/feedstock_root/build_artifacts/cffi_1696001742886/work
cfgv==3.4.0
charset-normalizer @ file:///home/conda/feedstock_root/build_artifacts/charset-normalizer_1698833585322/work
click==8.1.7
cloudpickle==3.0.0
colorama @ file:///home/conda/feedstock_root/build_artifacts/colorama_1666700638685/work
comm @ file:///home/conda/feedstock_root/build_artifacts/comm_1704278392174/work
contextlib2==0.5.5
cycler @ file:///home/conda/feedstock_root/build_artifacts/cycler_1696677705766/work
Cython @ file:///home/sat_bot/base/conda-bld/cython_1695764990393/work
debugpy @ file:///home/conda/feedstock_root/build_artifacts/debugpy_1695534305529/work
decorator @ file:///home/conda/feedstock_root/build_artifacts/decorator_1641555617451/work
defusedxml @ file:///home/conda/feedstock_root/build_artifacts/defusedxml_1615232257335/work
Deprecated @ file:///home/conda/feedstock_root/build_artifacts/deprecated_1685233314779/work
dgl==0.9.1
dgllife==0.3.2
distlib==0.3.7
docstring-parser==0.15
dpctl==0.15.0+42.g2a2c98f1f
dpnp==0.13.0+170.gf6d175b0f
e3nn==0.5.1
einops==0.7.0
emmet-core==0.64.0
entrypoints @ file:///home/conda/feedstock_root/build_artifacts/entrypoints_1643888246732/work
exceptiongroup @ file:///home/conda/feedstock_root/build_artifacts/exceptiongroup_1704921103267/work
executing @ file:///home/conda/feedstock_root/build_artifacts/executing_1698579936712/work
fastjsonschema @ file:///home/conda/feedstock_root/build_artifacts/python-fastjsonschema_1703780968325/work/dist
filelock==3.13.1
flake8==6.1.0
flake8-bandit==4.1.1
flake8-black==0.3.6
Flake8-pyproject==1.2.3
fonttools==4.25.0
frozenlist==1.4.0
fsspec==2023.10.0
future==0.18.3
geometric-algebra-attention==0.5.1
gitdb==4.0.11
GitPython==3.1.40
google-auth==2.26.2
google-auth-oauthlib==1.2.0
greenlet==3.0.3
grpcio==1.60.0
hyperopt==0.2.7
identify==2.5.32
idna @ file:///home/conda/feedstock_root/build_artifacts/idna_1663625384323/work
importlib-metadata==7.0.1
importlib-resources @ file:///home/conda/feedstock_root/build_artifacts/importlib_resources_1699364556997/work
iniconfig==2.0.0
intel-extension-for-pytorch @ file:///home/gta/workspace/LKG/build_conda_package_pytorch/binaries/intel_extension_for_pytorch-2.0.100%252Bcpu-cp39-cp39-linux_x86_64.whl#sha256=901fd4612125a1962621c7a5c9fced997723662761297987d9ab22a371208a11
ipykernel @ file:///home/conda/feedstock_root/build_artifacts/ipykernel_1707182759703/work
ipython @ file:///home/conda/feedstock_root/build_artifacts/ipython_1701831663892/work
ipython-genutils==0.2.0
ipywidgets @ file:///home/conda/feedstock_root/build_artifacts/ipywidgets_1694607144474/work
jedi @ file:///home/conda/feedstock_root/build_artifacts/jedi_1696326070614/work
Jinja2 @ file:///tmp/build/80754af9/jinja2_1624781299557/work
joblib @ file:///home/conda/feedstock_root/build_artifacts/joblib_1691577114857/work
jsonargparse==4.27.1
jsonschema @ file:///home/conda/feedstock_root/build_artifacts/jsonschema-meta_1705707496704/work
jsonschema-specifications @ file:///tmp/tmpkv1z7p57/src
jupyter @ file:///home/conda/feedstock_root/build_artifacts/jupyter_1696255489086/work
jupyter-client @ file:///home/conda/feedstock_root/build_artifacts/jupyter_client_1654730843242/work
jupyter-console @ file:///home/conda/feedstock_root/build_artifacts/jupyter_console_1678118109161/work
jupyter-server @ file:///home/conda/feedstock_root/build_artifacts/jupyter_server_1693923066986/work
jupyter_core @ file:///home/conda/feedstock_root/build_artifacts/jupyter_core_1704727023078/work
jupyterlab-widgets @ file:///home/conda/feedstock_root/build_artifacts/jupyterlab_widgets_1694598704522/work
jupyterlab_pygments @ file:///home/conda/feedstock_root/build_artifacts/jupyterlab_pygments_1707149102966/work
kiwisolver @ file:///home/conda/feedstock_root/build_artifacts/kiwisolver_1695379916629/work
latexcodec==2.0.1
lightning-utilities==0.10.0
llvmlite==0.41.1
lmdb==1.3.0
Markdown==3.5.2
markdown-it-py==3.0.0
MarkupSafe==2.1.3
matgl==0.9.2
matplotlib @ file:///tmp/build/80754af9/matplotlib-suite_1634667019719/work
matplotlib-inline @ file:///home/conda/feedstock_root/build_artifacts/matplotlib-inline_1660814786464/work
-e git+ssh://git@github.com/laserkelvin/matsciml.git@bbac579af862787ec635c98e2d18d596f2c0023c#egg=matsciml
mccabe==0.7.0
mdurl==0.1.2
mendeleev==0.14.0
mistune @ file:///home/conda/feedstock_root/build_artifacts/mistune_1698947099619/work
mkl-fft==1.3.6
mkl-random==1.2.2
mkl-service==2.4.0
mkl-umath==0.1.1
monty==2023.11.3
mp-api==0.33.3
mpmath @ file:///home/conda/feedstock_root/build_artifacts/mpmath_1678228039184/work
msgpack==1.0.7
multidict==6.0.4
munch==2.5.0
munkres==1.1.4
mypy-extensions==1.0.0
nbclassic @ file:///home/conda/feedstock_root/build_artifacts/nbclassic_1683202081046/work
nbclient @ file:///home/conda/feedstock_root/build_artifacts/nbclient_1684790896106/work
nbconvert @ file:///home/conda/feedstock_root/build_artifacts/nbconvert-meta_1707182911809/work
nbformat @ file:///home/conda/feedstock_root/build_artifacts/nbformat_1690814868471/work
nest_asyncio @ file:///home/conda/feedstock_root/build_artifacts/nest-asyncio_1705850609492/work
networkx @ file:///tmp/build/80754af9/networkx_1627459939258/work
neural-compressor @ file:///root/w0/workspace/lpot-nightly-release-wheel-build/lpot-models/dist/neural_compressor-2.3.1-py3-none-any.whl#sha256=d94ba25aad77289c24d475a1967889832a9712ba685afa03093a06d93b84d1d0
nodeenv==1.8.0
notebook @ file:///home/conda/feedstock_root/build_artifacts/notebook_1695225629675/work
notebook_shim @ file:///home/conda/feedstock_root/build_artifacts/notebook-shim_1682360583588/work
numba==0.58.1
numpy==1.26.2
oauthlib==3.2.2
oneccl-bind-pt @ file:///home/gta/workspace/LKG/build_conda_package_pytorch/binaries/oneccl_bind_pt-2.0.0%252Bcpu-cp39-cp39-linux_x86_64.whl#sha256=165c2d5525cf18390f8636fb18ee174b74ad7e785de5a000a9c63008e50f2182
opencv-python-headless @ file:///workspace/jsasswis/build_opecv/build_conda/opencv_python_headless-4.8.0.74-cp39-cp39-linux_x86_64.whl#sha256=aa6b111915aa44338293aa56605dd3d6e71998ba18b780d4907f0029b46a8fbb
opt-einsum==3.3.0
opt-einsum-fx==0.1.4
packaging @ file:///home/conda/feedstock_root/build_artifacts/packaging_1696202382185/work
palettable==3.3.3
pandas==1.5.3
pandocfilters @ file:///home/conda/feedstock_root/build_artifacts/pandocfilters_1631603243851/work
parso @ file:///home/conda/feedstock_root/build_artifacts/parso_1638334955874/work
pathspec==0.11.2
pbr==6.0.0
pexpect @ file:///home/conda/feedstock_root/build_artifacts/pexpect_1706113125309/work
pickleshare @ file:///home/conda/feedstock_root/build_artifacts/pickleshare_1602536217715/work
Pillow @ file:///home/conda/feedstock_root/build_artifacts/pillow_1695247205741/work
pkgutil_resolve_name @ file:///home/conda/feedstock_root/build_artifacts/pkgutil-resolve-name_1694617248815/work
platformdirs @ file:///home/conda/feedstock_root/build_artifacts/platformdirs_1696272223550/work
plotly==5.18.0
pluggy==1.3.0
pooch @ file:///home/conda/feedstock_root/build_artifacts/pooch_1698245576425/work
pre-commit==3.5.0
prettytable==0.7.2
prometheus-client @ file:///home/conda/feedstock_root/build_artifacts/prometheus_client_1700579315247/work
prompt-toolkit @ file:///home/conda/feedstock_root/build_artifacts/prompt-toolkit_1702399386289/work
protobuf==4.23.4
psutil @ file:///tmp/build/80754af9/psutil_1612297992929/work
ptyprocess @ file:///home/conda/feedstock_root/build_artifacts/ptyprocess_1609419310487/work/dist/ptyprocess-0.7.0-py2.py3-none-any.whl
pure-eval @ file:///home/conda/feedstock_root/build_artifacts/pure_eval_1642875951954/work
py-cpuinfo @ file:///tmp/build/80754af9/py-cpuinfo_1618437357364/work
py4j==0.10.9.7
pyasn1==0.5.1
pyasn1-modules==0.3.0
pybtex==0.24.0
pycocotools @ file:///home/conda/feedstock_root/build_artifacts/pycocotools_1626785515988/work
pycodestyle==2.11.1
pycparser @ file:///home/conda/feedstock_root/build_artifacts/pycparser_1636257122734/work
pydantic==1.10.12
pydantic-settings==2.1.0
pydantic_core==2.14.5
pyfiglet==0.8.post1
pyflakes==3.1.0
Pygments @ file:///home/conda/feedstock_root/build_artifacts/pygments_1700607939962/work
pymatgen==2023.7.20
pyparsing @ file:///home/conda/feedstock_root/build_artifacts/pyparsing_1690737849915/work
PySocks @ file:///home/conda/feedstock_root/build_artifacts/pysocks_1661604839144/work
pytest==7.4.3
python-dateutil==2.8.2
python-dotenv==1.0.0
pytorch-lightning==1.8.6
pytz @ file:///home/conda/feedstock_root/build_artifacts/pytz_1693930252784/work
PyYAML @ file:///home/conda/feedstock_root/build_artifacts/pyyaml_1695373447169/work
pyzmq @ file:///home/conda/feedstock_root/build_artifacts/pyzmq_1666828545060/work
qtconsole @ file:///home/conda/feedstock_root/build_artifacts/qtconsole-base_1700168901209/work
QtPy @ file:///home/conda/feedstock_root/build_artifacts/qtpy_1698112029416/work
rdkit==2023.3.1
referencing @ file:///home/conda/feedstock_root/build_artifacts/referencing_1706711412823/work
requests @ file:///home/conda/feedstock_root/build_artifacts/requests_1684774241324/work
requests-oauthlib==1.3.1
rich==13.7.0
rpds-py @ file:///home/conda/feedstock_root/build_artifacts/rpds-py_1705159800573/work
rsa==4.9
ruamel.yaml==0.18.5
ruamel.yaml.clib==0.2.8
ruff==0.2.1
schema==0.7.5
scikit-learn @ file:///home/conda/feedstock_root/build_artifacts/scikit-learn_1696574864087/work
scipy==1.10.1
Send2Trash @ file:///home/conda/feedstock_root/build_artifacts/send2trash_1682601222253/work
six @ file:///tmp/build/80754af9/six_1644875935023/work
smmap==5.0.1
sniffio @ file:///home/conda/feedstock_root/build_artifacts/sniffio_1662051266223/work
soupsieve @ file:///home/conda/feedstock_root/build_artifacts/soupsieve_1693929250441/work
spglib==2.1.0
SQLAlchemy==2.0.25
stack-data @ file:///home/conda/feedstock_root/build_artifacts/stack_data_1669632077133/work
stevedore==5.1.0
sympy @ file:///home/conda/feedstock_root/build_artifacts/sympy_1684180539862/work
tabulate==0.9.0
TBB==0.2
tenacity==8.2.3
tensorboard==2.15.1
tensorboard-data-server==0.7.2
tensorboardX==2.6.2.2
terminado @ file:///home/conda/feedstock_root/build_artifacts/terminado_1699810101464/work
threadpoolctl @ file:///home/conda/feedstock_root/build_artifacts/threadpoolctl_1689261241048/work
tinycss2 @ file:///home/conda/feedstock_root/build_artifacts/tinycss2_1666100256010/work
tomli==2.0.1
torch @ file:///home/gta/workspace/LKG/build_conda_package_pytorch/binaries/torch-2.0.1%252Bcpu-cp39-cp39-linux_x86_64.whl#sha256=73482a223d577407c45685fde9d2a74ba42f0d8d9f6e1e95c08071dc55c47d7b
torch-scatter==2.1.2
torch-sparse==0.6.18
torch_geometric==2.4.0
torchmetrics==1.2.1
torchvision @ file:///home/gta/workspace/LKG/build_conda_package_pytorch/binaries/torchvision-0.15.2a0%2Bfa99a53-cp39-cp39-linux_x86_64.whl#sha256=9051b9e66fca6dcc8ef5118adb9ddebb28dab1fe966e9baffefdb2eba183be52
tornado @ file:///tmp/build/80754af9/tornado_1606942317143/work
tqdm==4.66.1
traitlets @ file:///home/conda/feedstock_root/build_artifacts/traitlets_1704212992681/work
typeshed-client==2.4.0
typing_extensions @ file:///home/conda/feedstock_root/build_artifacts/typing_extensions_1695040754690/work
uncertainties==3.1.7
urllib3 @ file:///home/conda/feedstock_root/build_artifacts/urllib3_1697720414277/work
virtualenv==20.25.0
wcwidth @ file:///home/conda/feedstock_root/build_artifacts/wcwidth_1704731205417/work
webencodings @ file:///home/conda/feedstock_root/build_artifacts/webencodings_1694681268211/work
websocket-client @ file:///home/conda/feedstock_root/build_artifacts/websocket-client_1701630677416/work
Werkzeug==3.0.1
widgetsnbextension @ file:///home/conda/feedstock_root/build_artifacts/widgetsnbextension_1694598693908/work
wrapt @ file:///tmp/build/80754af9/wrapt_1638433857881/work
yarl==1.9.3
zipp @ file:///home/conda/feedstock_root/build_artifacts/zipp_1695255097490/work
```

--- a/matsciml/models/pyg/mace/tools/scatter.py
+++ b/matsciml/models/pyg/mace/tools/scatter.py
@@ -16,6 +16,8 @@ See https://github.com/pytorch/pytorch/issues/63780.
 
 from __future__ import annotations
 
+from typing import Optional
+
 import torch
 
 
@@ -36,8 +38,8 @@ def scatter_sum(
     src: torch.Tensor,
     index: torch.Tensor,
     dim: int = -1,
-    out: torch.Tensor | None = None,
-    dim_size: int | None = None,
+    out: Optional[torch.Tensor] = None,
+    dim_size: Optional[int] = None,
     reduce: str = "sum",
 ) -> torch.Tensor:
     assert reduce == "sum"  # for now, TODO
@@ -61,8 +63,8 @@ def scatter_std(
     src: torch.Tensor,
     index: torch.Tensor,
     dim: int = -1,
-    out: torch.Tensor | None = None,
-    dim_size: int | None = None,
+    out: Optional[torch.Tensor] = None,
+    dim_size: Optional[int] = None,
     unbiased: bool = True,
 ) -> torch.Tensor:
     if out is not None:
@@ -99,8 +101,8 @@ def scatter_mean(
     src: torch.Tensor,
     index: torch.Tensor,
     dim: int = -1,
-    out: torch.Tensor | None = None,
-    dim_size: int | None = None,
+    out: Optional[torch.Tensor] = None,
+    dim_size: Optional[int] = None,
 ) -> torch.Tensor:
     out = scatter_sum(src, index, dim, out, dim_size)
     dim_size = out.size(dim)

